File: Boot/App/src/main/java/org/elkoserver/foundation/boot/BootArguments.java
package org.elkoserver.foundation.boot;

import org.elkoserver.foundation.properties.ElkoProperties;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class BootArguments {

    final String mainClassName;
    final ElkoProperties bootProperties;

    BootArguments(String[] args) {
        super();

        List<String> remainingArgs = new ArrayList<>(args.length);
        Map<String, String> props = new HashMap<>(args.length);
        scanArgsForProperties(args, props, remainingArgs);

        if (remainingArgs.size() < 1) {
            throw new Error("Boot needs class name to boot");
        }

        mainClassName = remainingArgs.get(0);
        bootProperties = new ElkoProperties(props);
    }

    private void scanArgsForProperties(String[] args, Map<String, String> props, List<String> remainingArgs) {
        for (String arg : args) {
            int j = arg.indexOf('=');
            if (arg.indexOf('=') > 0) {
                String key = arg.substring(0, j);
                String value = arg.substring(j + 1);
                props.put(key, value);
            } else {
                remainingArgs.add(arg);
            }
        }
    }
}

File: Boot/App/src/main/kotlin/org/elkoserver/foundation/boot/BootArguments.kt
package org.elkoserver.foundation.boot

import org.elkoserver.foundation.properties.ElkoProperties

internal class BootArguments internal constructor(vararg arguments: String) {
    internal val mainClassName: String

    internal val bootProperties: ElkoProperties

    init {
        val argumentsThatAreNotProperties = ArrayList<String>(arguments.size)
        val properties = HashMap<String, String>(arguments.size)
        scanArgumentsForProperties(arguments, properties, argumentsThatAreNotProperties)
        require(1 <= argumentsThatAreNotProperties.size) { "Boot needs class name to boot" }
        mainClassName = argumentsThatAreNotProperties[0]
        bootProperties = ElkoProperties(properties)
    }

    private fun scanArgumentsForProperties(arguments: Array<out String>, destinationProperties: MutableMap<String, String>, destinationArgumentsThatAreNotProperties: MutableList<String>) {
        arguments.forEach { argument ->
            scanArgumentForProperty(argument, destinationProperties, destinationArgumentsThatAreNotProperties)
        }
    }

    private fun scanArgumentForProperty(argument: String, destinationProperties: MutableMap<String, String>, destinationArgumentsThatAreNotProperties: MutableList<String>) {
        val indexOfEqualsSign = argument.indexOf('=')
        if (0 < indexOfEqualsSign) {
            val key = argument.take(indexOfEqualsSign)
            val value = argument.substring(indexOfEqualsSign + 1)
            destinationProperties[key] = value
        } else {
            destinationArgumentsThatAreNotProperties.add(argument)
        }
    }
}
